### ISSUE TITLE
Low Queue Activity - Packet Loss

### ISSUE THREAD

**Issue by Fr3DBr** [luigirizzo/netmap#758]
Currently, I'm using NIOCTXSYNC/NIOCRXSYNC to move/sync packets, I've noticed, that, when there is a too small activity, a certain packet loss will happen.

My Network Driver is ixgbe (netmap patched) version 5.6.5. (I've tried newer and older drivers, but the result is the same so far).

By "too small activity" I mean like having  eg: 1 packet per second, hitting one of the queues, there will be packet loss like this:

Reply from 10.230.0.1: bytes=32 time=2678ms TTL=62
Reply from 10.230.0.1: bytes=32 time=1507ms TTL=62
Reply from 10.230.0.1: bytes=32 time=1487ms TTL=62
Request timed out.
Reply from 10.230.0.1: bytes=32 time=1195ms TTL=62
Reply from 10.230.0.1: bytes=32 time=563ms TTL=62
Reply from 10.230.0.1: bytes=32 time=2198ms TTL=62
Reply from 10.230.0.1: bytes=32 time=579ms TTL=62
Request timed out.
Reply from 10.230.0.1: bytes=32 time=1451ms TTL=62
Reply from 10.230.0.1: bytes=32 time=307ms TTL=62
Request timed out.
Request timed out.

Although, as soon this packet rate increases to eg: 5~10 packets/second (in the same packet queue as before), this problem is gone, like here:

Pinging 10.230.0.1 with 32 bytes of data:
Reply from 10.230.0.1: bytes=32 time=5ms TTL=62
Reply from 10.230.0.1: bytes=32 time=5ms TTL=62
Reply from 10.230.0.1: bytes=32 time=8ms TTL=62
Reply from 10.230.0.1: bytes=32 time=5ms TTL=62
Reply from 10.230.0.1: bytes=32 time=6ms TTL=62
Reply from 10.230.0.1: bytes=32 time=8ms TTL=62
Reply from 10.230.0.1: bytes=32 time=7ms TTL=62
Reply from 10.230.0.1: bytes=32 time=6ms TTL=62
Reply from 10.230.0.1: bytes=32 time=6ms TTL=62

Basically this means, I'd need at least 5 packets per second per queue (5 * 8 queues = 40 pps in my case), to avoid this strange issue.

Any hints of why this happens, and if there's a way to avoid it ? If we do it, by using BUSYWAIT, the problem is gone, but I like more, the idea of saving CPU cycles by using the ioctl calls. :)

For me, it looks like there's some "wait timeout" inside these ioctl calls ? If yes, can we tune/adjust this somehow ?

Not that this is a "big issue", although, it causes some troubles with monitoring, so if this could be handled, it would be great.

**Comment by Fr3DBr** [comment]
I'm wondering if anyone else has noticed this behavior so far ?

**Comment by Fr3DBr** [comment]
I had the following in my interface loop:

--

                /* Check for pending packets */
                int n0 = pkt_queued(pTI->pb, 0); // NIC [RX]
                int n1 = pkt_queued(pTI->pb, 1); // NIC [TX]

                if (n0) { // NIC RX -> HOSTSTACK TX SYNC
                        ioctl(pollfd[0].fd, NIOCTXSYNC, NULL);
                        pollfd[0].revents = POLLOUT;
                } else {
                        ioctl(pollfd[1].fd, NIOCRXSYNC, NULL);
                }
                if (n1) { // FLUSHING NIC TX - IF NEEDED
                        ioctl(pollfd[1].fd, NIOCTXSYNC, NULL);
                }

                // NIC[RX] -> HOST STACK[TX]
                if (verbose && pollfd[1].revents & POLLERR) {
                        struct netmap_ring *rx = pTI->pb_rxring;
                        D("{%d} error on fd1, rx [%d,%d,%d)", pTI->ring_id, rx->head, rx->cur, rx->tail);
                        pTI->poll_skip_count++;
                }
                if (pollfd[0].revents & POLLOUT) {
                        process_rings(pTI->pb_rxring, pTI->pa_txring, 1, pTI, nCurTime, nCurTick);
                }

--

It works fine, but the problem above happens... :/

I could only "fix this" by using a hack this way:

--

                /* Check for pending packets */
                int n0 = pkt_queued(pTI->pb, 0); // NIC [RX]
                int n1 = pkt_queued(pTI->pb, 1); // NIC [TX]

                if (n0) { // NIC RX -> HOSTSTACK TX SYNC
                        ioctl(pollfd[0].fd, NIOCTXSYNC, NULL);
                        pollfd[0].revents = POLLOUT;
                } else {
                        /* Busywait Hack */
                        if ( nPacketPerSec <= 8 ) { <<<<<<
                                ioctl(pollfd[0].fd, NIOCTXSYNC, NULL); <<<<<
                                pollfd[0].revents = POLLOUT; <<<<<
                        } <<<<<
                        ioctl(pollfd[1].fd, NIOCRXSYNC, NULL);
                }
                if (n1) { // FLUSHING NIC TX - IF NEEDED
                        ioctl(pollfd[1].fd, NIOCTXSYNC, NULL);
                }

                // NIC[RX] -> HOST STACK[TX]
                if (verbose && pollfd[1].revents & POLLERR) {
                        struct netmap_ring *rx = pTI->pb_rxring;
                        D("{%d} error on fd1, rx [%d,%d,%d)", pTI->ring_id, rx->head, rx->cur, rx->tail);
                        pTI->poll_skip_count++;
                }
                if (pollfd[0].revents & POLLOUT) {
                        process_rings(pTI->pb_rxring, pTI->pa_txring, 1, pTI, nCurTime, nCurTick);
                }

--

Then, there is no more packet loss at small loads, but I suspect, there might be a better/different way of doing the same ? :O

**Comment by Fr3DBr** [comment]
Hi, guys. Any light on this issue ? :(

**Comment by jhk098** [comment]
It looks like you are trying to bridge a NIC to its host rings, in the same way the netmap `bridge` application does, e.g.:
````
# ./bridge -i ens4
````
Does the issue show up with `bridge`?

**Comment by Fr3DBr** [comment]
Hi,

It does not happen with "poll()" mode bridge, however when we are using bridge-b (note the b at the end), this one, is the BUSYWAIT mode enabled version of Bridge App, then yes, the problem can be reproduced there as well.

**Comment by jhk098** [comment]
what happens if you apply the following patch?
````
diff --git a/apps/bridge/bridge.c b/apps/bridge/bridge.c
index 77d235bf..b5b0663e 100644
--- a/apps/bridge/bridge.c
+++ b/apps/bridge/bridge.c
@@ -323,13 +323,11 @@ main(int argc, char **argv)
 		n1 = rx_slots_avail(pb);
 #if defined(_WIN32) || defined(BUSYWAIT)
 		if (n0) {
-			ioctl(pollfd[1].fd, NIOCTXSYNC, NULL);
 			pollfd[1].revents = POLLOUT;
 		} else {
 			ioctl(pollfd[0].fd, NIOCRXSYNC, NULL);
 		}
 		if (n1) {
-			ioctl(pollfd[0].fd, NIOCTXSYNC, NULL);
 			pollfd[0].revents = POLLOUT;
 		} else {
 			ioctl(pollfd[1].fd, NIOCRXSYNC, NULL);
@@ -375,11 +373,19 @@ main(int argc, char **argv)
 			D("error on fd1, rx [%d,%d,%d)",
 			    rx->head, rx->cur, rx->tail);
 		}
-		if (pollfd[0].revents & POLLOUT)
+		if (pollfd[0].revents & POLLOUT) {
 			ports_move(pb, pa, burst, msg_b2a);
+#if defined(_WIN32) || defined(BUSYWAIT)
+			ioctl(pollfd[0].fd, NIOCTXSYNC, NULL);
+#endif
+		}
 
-		if (pollfd[1].revents & POLLOUT)
+		if (pollfd[1].revents & POLLOUT) {
 			ports_move(pa, pb, burst, msg_a2b);
+#if defined(_WIN32) || defined(BUSYWAIT)
+			ioctl(pollfd[1].fd, NIOCTXSYNC, NULL);
+#endif
+		}
 
 		/*
 		 * We don't need ioctl(NIOCTXSYNC) on the two file descriptors.

````

**Comment by Fr3DBr** [comment]
Your patch works, the problem is fixed this way. The interesting point, is now that the CPU is not at 100% anymore. It's 0% when idle, and working great so far.
